akka-typed-tests/src/test/scala/akka/typed/scaladsl/persistence/ApiTest.scala

```diff
@@ -0,0 +1,239 @@
+package akka.typed.scaladsl.persistence
+
+import akka.typed
+
+import scala.concurrent.duration._
+import akka.typed.{ ActorRef, Behavior, ExtensibleBehavior, Signal }
+import akka.typed.scaladsl.{ ActorContext, TimerScheduler }
+
+import scala.concurrent.ExecutionContext
+
+class ApiTest {
+  object TypedPersistentActor {
+
+    sealed trait PersistentEffect[+Event]
+
+    case object PersistNothing extends PersistentEffect[Nothing]
+
+    case class PersistAnd[Event](event: Event, callback: Event ⇒ Unit) extends PersistentEffect[Event]
+
+    case class Persist[Event](event: Event) extends PersistentEffect[Event] {
+      def andThen(callback: Event ⇒ Unit) = PersistAnd(event, callback)
+    }
+
+  }
+
+  object Actor {
+    import TypedPersistentActor._
+
+    class PersistentBehavior[Command, Event, State] extends ExtensibleBehavior[Command] {
+      override def receiveSignal(ctx: typed.ActorContext[Command], msg: Signal): Behavior[Command] = ???
+      override def receiveMessage(ctx: typed.ActorContext[Command], msg: Command): Behavior[Command] = ???
+
+      def onRecoveryComplete(callback: (ActorContext[Command], State) ⇒ Unit): PersistentBehavior[Command, Event, State] = ???
+      def snapshot[Snapshot](
+        onState: State ⇒ Option[Snapshot]          = (_: State) ⇒ None,
+        onEvent: Event ⇒ Option[Snapshot]          = (_: Event) ⇒ None,
+        on:      (State, Event) ⇒ Option[Snapshot] = (_: State, _: Event) ⇒ None,
+        recover: Snapshot ⇒ Option[State]
```

Comment:
what is `Snapshot` vs. `State`? isn't the type of the stored snapshot always `State`?

Reply:
My thinking here was that `State` might be (entirely) serializable: for example the `ExplicitSnapshots` example below only takes snapshot when there are not tasks in flight, meaning tasks don't need to be serializable.

That example is a bit contrived because the tasks are serialized in the events anyway, but in general I've found it to be helpful to separate the events and snapshots (which are persistent and can be tricky to evolve) from the commands and state (which are transient and can therefore be refactored freely).

Reply:
I find that complicated. Shouldn't the serializer know how to ignore transient fields if any.

Reply:
That might be dangerous though: developers might expect certain fields to survive serialization while they don't. I like how refactoring `State` for 'logic-only' changes would only affect the conversions in `onState`/`recover`, and not the serialization itself. Maybe it'd be good to get some data on how snapshotting is used in practice?

Reply:
The snapshot must anyway be a subset of the state and it must be possible to create the state from the snapshot. That only leaves room for transient fields (e.g. a cache), and if you don't want to handle that in the serializer I think it's more like `state.clearForSnapshot()` before you save it. And it must anyway immutable.

I see your point in having different classes because you want to evolve them differently, i.e. this is a mapping layer. Not sure we should have yet another way of mapping things for evolving. We already have guidelines for how to do that with serialization (and event adapaters).

If going down this path I think it should be a pure mapping function `Snapshot => State`, `State => Snapshot` instead of convoluting it with **when** the snapshot is to be taken.

Reply:
Another parameter, I think we should also support automatic snapshotting after a given number of events. That should be the normal use of snapshots, a plain optimization.